src/mainloop.rkt

```diff
@@ -168,36 +166,40 @@
 (define (reconstruct! alts)
   ;; extracts the base expression of a patch
   (define (get-starting-expr altn)
-    (match altn
-     [(alt prog '(patch) _) (program-body prog)]
-     [(alt _ _ (list)) #f]
-     [(alt _ _ (list prev)) (get-starting-expr prev)]))
+    (match (alt-event altn)
+     ['(patch) (alt-expr altn)]
+     [_
+      (if (null? (alt-prevs altn))
+          #f
+          (get-starting-expr (first (alt-prevs altn))))]))
 
   ;; takes a patch and converts it to a full alt
   (define (reconstruct-alt altn loc0 orig)
     (let loop ([altn altn])
-      (match-define (alt prog event prev) altn)
+      (match-define (alt _ event prevs) altn)
       (cond
        [(equal? event '(patch)) orig]
        [else
         (define event*
+          ;; The 2 at the start of locs is left over from when we
+          ;; differentiated between "programs" with a λ term and
+          ;; "expressions" without
           (match event
-           [(list 'taylor name var loc)
-            (list 'taylor name var (append loc0 (cdr loc)))]
-           ; TODO : Recosnsruct and revise history
-           [`(rr, loc, input, proof, soundiness)
-            (list 'rr (append loc0 (cdr loc)) input proof soundiness)]
-           [`(simplify ,loc ,input ,proof ,soundiness)
-            (list 'simplify (append loc0 (cdr loc)) input proof soundiness)]))
-        (define prog* (location-do loc0 (alt-program orig) (λ (_) (program-body prog))))
-        (alt prog* event* (list (loop (first prev))))])))
+           [(list 'taylor name var (list 2 loc ...))
+            (list 'taylor name var (append '(2) loc0 loc))]
+           [(list 'rr (list 2 loc ...) input proof soundiness)
+            (list 'rr (append '(2) loc0 loc) input proof soundiness)]
+           [(list 'simplify (list 2 loc ...) input proof soundiness)
+            (list 'simplify (append '(2) loc0 loc) input proof soundiness)]))
```

Comment:
I think the `2` should be removed from any mainloop code. Sure, it's there for historical reasons, but it doesn't really make sense here. Might need to be prepended in the report generation. Doesn't have to be this PR.